### DISCUSSION TITLE
"NULL value passed as symbol address" in segment_trees

### DISCUSSION THREAD

**Discussion by searoy** [r-lidar/lidR#547] [Q&A] [answered]
As the title says, I am getting a NULL error in my code below.

```r
# Read dtm for normalization
dtm <- raster("AlexSub2DTM.tif")

# Read trunks
trunks <- readOGR(dsn = "trunksfilterout2o.shp")

# read Large las as lasCatalog
las1<- readLAScatalog("Alex2swathclipped.las")

# Variables for processing
opt_output_files(las1) <- "./lasCat/laschunk_{ID}"
opt_chunk_size(las1) <- 495 # Processing chunks of 495 x 495
opt_chunk_buffer(las1) <- 15 # 15 ft buffer
opt_stop_early(las1) <- TRUE # or FALSE
opt_filter(las1) <- "-drop_z_below 15"

# Normalize
las1 <- normalize_height(las1, dtm,
                         na.rm = TRUE, 
                        method = "bilinear",
                        overwrite = TRUE)

# CHM
opt_output_files(las1) <- "./lasCat/laschml_{ID}"
chml1 <- rasterize_canopy(las1, 0.25,
                          pitfree(c(0,2,5,10), c(0,1),
                                  subcircle = 0.2),
                          overwrite = TRUE)

# Smoothing
ker = matrix(1,3,3)
chml1 = focal(chml1, w = ker, fun = median)
chml1 = focal(chml1, w = ker, fun = median)

# Save output
writeRaster(chml1, filename="As2_chml3_pf.tif", overwrite=TRUE)

# Tree Segmentation
algotr1 <- silva2016(chml1, trunks, exclusion = 0.5, ID = "treeID")
lastr1 <- segment_trees(las1, algotr1)
```

This is a small subset from a much larger project so scalability is non-negotiable and clipping is the smartest way for me to reduce total file size. My las is a non-rectangular ROI clip (pit-free and no voids/holes). 
Would this cause the NULL error or should I look elsewhere?

(zip with all data is over 1Gb and won't upload)

```r
lastr1 <- segment_trees(las1, algotr1)
#> Processing [======>-----------------------------------------------------------]  8% (1/12) eta:  2m
#> Error: NULL value passed as symbol address
```

**Comment by Jean-Romain** [comment]
> Error: NULL value passed as symbol address

Is an `Rcpp` error. An address to an non-existing object has been passed to C++ code... It's all I can say without a reproducible example. I suspect `terra` being involved somewhere but I can't be sure.

Please open an issue with a minimal reproducible example. A minimal reproducible example should look like that I guess. And if it fails at the first file I guess you only need to use one LAS file to reproduce so you can share it + the rasters

```r
library(llidR)
library(terra)
library(sf)

chm1 <- rast("As2_chml3_pf.tif")
trunk <- st_read("????")
nlas <- readLAScatalog("./lasCat/")

# Tree Segmentation
algotr1 <- silva2016(chm1, trunks, exclusion = 0.5, ID = "treeID")
lastr1 <- segment_trees(nlas1, algotr1)
```

**Comment by searoy** [comment]
This is a "microset" of my subset (one of the 495x495 chunks from the lasCAT). The total dataset is xTB and growing. As you see it is a nonrectangular clip using a variable radius buffer around trees in close proximity to powerlines. I suspect `rasterize_canopy` is expecting a rectangular grid/raster and that this odd-shaped raster is the root of the issue, but I don't know enough to know. I ran this microset and it fails the same as the large lasCAT nonrectangular clip. When I run the rectangular unclipped subset it succeeds without issue.

The trunk shapefile replaces the tree top input since trunks move less across multiple collects.
USGS dtm is used for normalization to save processing time and standardize ground across collects.

[Searoy_las.zip](https://github.com/r-lidar/lidR/files/8245169/Searoy_las.zip)

**Reply by Jean-Romain** [reply]
Please share R code to reproduce

**Reply by searoy** [reply]
R code is above

**Reply by Jean-Romain** [reply]
I cannot reproduce. Please share a minimal reproducible example. It means a code chunk I can copy paste without changing a single letter and run to trigger the error. Your code is not reproducible. The errors I get are CRS errors because the different objects have different CRS. I fixed this and everything ran ok.

**Comment by searoy** [comment, marked as answer]
I don't think the issue I'm seeing is the fault of lidR. Due to the size of our data I'm trying to leverage hyperthreading, but it fails with `future` and `multisession `enabled. Without them it runs.

```
library(lidR)
library(terra)
library(raster)
library(rgdal)
library(sf)
library(future)

# system setup
plan(multisession, workers = 6L)
set_lidr_threads(30L)
```

**Reply by Jean-Romain** [reply]
According to your code you have 180 cores. Unless you really have 180 cores OpenMP will be deactivated. Am I supposed to run the previous code with 

```r
plan(multisession, workers = 2L)
set_lidr_threads(2L)
```

to reproduce?  (I only have 4 cores on my machine)

**Reply by searoy** [reply]
I have 24c/48t Threadripper. but even 2/2 as you have listed doesn't work. Disabling multisession does.

```
#plan(multisession, workers = 2L)
set_lidr_threads(24L)
```

This gives me 12 `Error in x$.self$finalize() : attempt to apply non-function` errors.

```
#plan(multisession, workers = 2L)
set_lidr_threads(12L)
```

This gives no errors.
No version of `plan(multisession)` worked for me.

TYVM for your help with this.

**Reply by Jean-Romain** [reply]
```
Error in x$.self$finalize() : attempt to apply non-function
```

Is a `terra` issue that the maintainer do not want to fix but we do not care of it it is not a real problem. This is not the error you initially reported. So, please, report something reproducible if you want me to help you.
